### DISCUSSION TITLE
Changing Environment on Runtime

### DISCUSSION THREAD

**Discussion by YoussefSalameh** [abpframework/abp#8255] [General]
I was using **dispatchSetEnvironment()** to change my environment on runtime but after upgrading I wasn't able to do that using this.environmentService.setState(env) because it will not change my environment variables that are already being stored at start.

@maliming  @hikalkan @realLiangshiwei 
Please can anyone provide us with a solution ?

**Comment by gterdem** [comment]
Can you try using `setState` method of new [EnvironmentService](https://docs.abp.io/en/abp/4.2/UI/Angular/Environment#environmentservice)?

**Reply by YoussefSalameh** [reply]
I've already tried this and it didn't work.

**Reply by mehmet-erim** [reply]
Hi @YoussefSalameh 

It works as expected. See the code below and result:

```ts
import { EnvironmentService } from '@abp/ng.core';

  constructor(private environment: EnvironmentService) {
    const env = this.environment.getEnvironment();
    console.log(env);
    
    this.environment.setState({ ...env, application: { ...env.application, name: 'NewAppName' } });
    console.warn(this.environment.getEnvironment());
  }
```


![image](https://user-images.githubusercontent.com/34455572/113893380-a0fea700-97cf-11eb-905a-57e6b0f66f24.png)

The problem may be related to your logic. If you share the code with me, I can help you.

**Reply by YoussefSalameh** [reply]
![Capture1](https://user-images.githubusercontent.com/63226854/114009511-e1136780-986b-11eb-8a70-0f9a8b670e7b.PNG)

Please Note that I want to change both apis and oauthConfig before log in to my application.

![Capture](https://user-images.githubusercontent.com/63226854/114009028-76fac280-986b-11eb-8085-8b5689c52c88.PNG)

Should I do something else after using setState() ?
And How I should handle it after logging in because it will return to the initial environment ?

**Reply by mehmet-erim** [reply]
- Did you see the [remote environment section in environment document](https://docs.abp.io/en/abp/latest/UI/Angular/Environment#remoteenvironment)? This may help you to achieve your goal.

- You can set the environment by getting app name from session storage on app initialization. (maybe you can use the APP_INITIALIZER token to do this)

- You can override the `setState` method of `EnvironmentState`. You can;
     * Create a service called `MyEnvironmentService`. 
     * Extend the `EnvironmentService`.
     * Override the `setState` method and add your logic here
     * Override the `EnvironmentService` with the new service like this:
    
    ```ts
    // app.module.ts
    
    //...
    
    providers: [{provide: EnvironmentService, useClass: MyEnvironmentService}]
    
    ```

**Reply by YoussefSalameh** [reply]
Dear @mehmet-erim 

Please Note That Remote Environment solved my issue.

Thanks
